### DISCUSSION TITLE
ConnectionError when running TigerGraph query

### DISCUSSION THREAD

**Discussion by xuetaoshi** [graphistry/graph-app-kit#90] [Q&A]
Hi,
I'm trying to setup a local web app connecting a local tigergraph server running on a private remote server. The same installed query can be run without problem using pyTigergraph on a Jupyter notebook (I'm using a SSH tunnel to forward port 9000 and 14240). But running the same query inside of the app view Python code produces ConnectionError. The full error message is attached below. I tried to add port forwarding in docker parameter files, docker-compose.yml under /src/docker/ and docker-compose.override.yml under /src/docker/override/. None of these worked so far. As I'm not that proficient at docker, I'd really appreciate if anyone can point me to where I should setup this docker network configuration and solve this error. Please let me know if there is any additional information I should provide. Thank you so much!

ConnectionError: HTTPConnectionPool(host='localhost', port=9000): Max retries exceeded with url: /query/protein_fragment/PDBKnowledgeGraph?p=7D2V (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0xffff75738040>: Failed to establish a new connection: [Errno 111] Connection refused'))
Traceback:

File "/apps/views/protein_fragment/__init__.py", line 511, in run_all
    filter_pipeline_result = run_filters(sidebar_filters['pdb_code'], conn)
File "/usr/local/lib/python3.8/site-packages/streamlit/runtime/legacy_caching/caching.py", line 618, in wrapped_func
    return get_or_create_cached_value()
File "/usr/local/lib/python3.8/site-packages/streamlit/runtime/legacy_caching/caching.py", line 600, in get_or_create_cached_value
    return_value = non_optional_func(*args, **kwargs)
File "/apps/views/protein_fragment/__init__.py", line 192, in run_filters
    results = conn.runInstalledQuery("PDBKnowledgeGraph", {"p": pdb_code})
File "/usr/local/lib/python3.8/site-packages/pyTigerGraph/pyTigerGraphQuery.py", line 218, in runInstalledQuery
    ret = self._get(self.restppUrl + "/query/" + self.graphname + "/" + queryName,
File "/usr/local/lib/python3.8/site-packages/pyTigerGraph/pyTigerGraphBase.py", line 318, in _get
    res = self._req("GET", url, authMode, headers, None, resKey, skipCheck, params, strictJson)
File "/usr/local/lib/python3.8/site-packages/pyTigerGraph/pyTigerGraphBase.py", line 272, in _req
    res = requests.request(method, url, headers=_headers, data=_data, params=params, verify=verify)
File "/usr/local/lib/python3.8/site-packages/requests/api.py", line 59, in request
    return session.request(method=method, url=url, **kwargs)
File "/usr/local/lib/python3.8/site-packages/requests/sessions.py", line 587, in request
    resp = self.send(prep, **send_kwargs)
File "/usr/local/lib/python3.8/site-packages/requests/sessions.py", line 701, in send
    r = adapter.send(request, **kwargs)
File "/usr/local/lib/python3.8/site-packages/requests/adapters.py", line 565, in send
    raise ConnectionError(e, request=request)

**Comment by webcoderz** [comment]
Prob have to open up the ports for the graphappkit & jupyter containers on the docker compose to the corresponding ports needed to expose the docker internal network to the host network on those ports so you can connect with pytigergraph to the remote server. (Docker internal networking is typically segregated from your hosts network)

**Reply by xuetaoshi** [reply]
Thanks for the fast reply! That confirms the direction I'm taking right now. But I'm not sure how to achieve that. So far I have added the following lines to /src/docker/Dockerfile
EXPOSE 8501
EXPOSE 9000
EXPOSE 14240
And added the following to the "ports:" section in  /src/docker/docker-compose.yml and /src/docker/override/docker-compose.override.yml 
      - "14022:22"
      - "9000:9000"
      - "14240:14240"
      
 Again I'm not very familiar with using docker so the above are what I got from Googling so these may seem pointless to experienced docker user. But I'd be really grateful if you can point me to where I should be looking at in order to connect the docker network with host network for running the tigergraph query. Thank you so much!

**Reply by webcoderz** [reply]
Looks good to me! check and let me know!

**Reply by xuetaoshi** [reply]
These are what I have tried so far and the error still persists. I even tried removing docker-compose rm -f and rebuild but still no luck. I'm using the cpu version. Could it be because the query call was made outside of streamlit somehow? The code was in the run_filters() functional call as in the demo_tigergraph_fraud and with the streamlit annotation. So I'm not sure what's wrong here.
